### ISSUE TITLE
An entry on stick-table disappears unintentionally

### ISSUE THREAD

**Issue by kyontan** [haproxy/haproxy#472]
Hello! I met a problem when I tried Active-Standby configuration using HAProxy.
If there might be a problem about HAProxy, please tell me how to solve issue.

## Output of `haproxy -vv` and `uname -a`

```
$ uname -a
Linux lb1 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ haproxy -vv
HA-Proxy version 2.1.0 2019/11/25 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.0.html
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO -OPENSSL -LUA +F
UTEX +ACCEPT4 -MY_ACCEPT4 -ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [TRACE] trace
        [COMP] compression
```

## What's the configuration?

```
global
    daemon
    log /dev/log local0 info
    stats socket /var/run/haproxy.sock mode 600 level admin
    stats timeout 10m

defaults
    log global
    mode tcp
    timeout connect 1000ms
    timeout client 1000ms
    timeout server 1000ms
    maxconn 32

frontend frontend_a
    bind *:1935
    default_backend servers_a

backend servers_a
    mode tcp
    stick-table type integer size 1
    stick on dst_port

    default-server check inter 1s fall 3 rise 3
    server server1 backend1:1935
    server backup backend2:1935 backup
```

## Steps to reproduce the behavior

1. Establish TCP connection to `1935/tcp` that HAProxy listens to
    - (SYN -> SYN/ACK -> ACK)
2. Disconnect TCP connection without any payload from client
    - (FIN -> FIN/ACK -> ACK)

## Actual behavior

- stick-table `servers_a` sometimes become empty though an entry is stored before becoming empty.

```
$ echo "show table servers_a" | sudo socat /var/run/haproxy.sock stdio
# table: servers_a, type: integer, size:1, used:1
0x7f6ccc017a90: key=1935 use=0 exp=0 server_id=1 server_name=server1
$ nc -z localhost 1935
$ echo "show table servers_a" | sudo socat /var/run/haproxy.sock stdio
# table: servers_a, type: integer, size:1, used:0
```

## Expected behavior

- stick-table `servers_a` is configured as a 1-entry table, so that, any established connection should keep a entry stored
  - In this case, HAProxy should retain entry that mean dst_port(`1935`) -> backend (`server1`)
  - There is no `nopurge` and no `expire` option, so that, table with a entry shouldn't get empty.

```
$ echo "show table servers_a" | sudo socat /var/run/haproxy.sock stdio
# table: servers_a, type: integer, size:1, used:1
0x7f6ccc017a90: key=1935 use=0 exp=0 server_id=1 server_name=server1
$ nc -z localhost 1935
$ echo "show table servers_a" | sudo socat /var/run/haproxy.sock stdio
# table: servers_a, type: integer, size:1, used:1
0x7f6ccc017a90: key=1935 use=0 exp=0 server_id=1 server_name=server1
```

Note that `nc -z` establishes connection then disconnects that without any data transfer.

```
$ nc -v
Ncat: Version 7.50 ( https://nmap.org/ncat )
Ncat: You must specify a host to connect to. QUITTING.
```

## Do you have any idea what may have caused this?

- I have noticed that this behavior can be observed 100% when I try `nc -z` from localhost of HAProxy, and an instance on same L2 network (in this case, I tested on AWS EC2).
- When I try BSD nc on macOS, entry had been kept (== cannot observed)
- When I try nmap-netcat on Docker on macOS, it **sometimes** observed.
- I think this might be something like race-conditional behavior

In my macOS environment:
```
$ uname -a
Darwin (hostname) 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
$ man nc
(snip)
AUTHORS
     Original implementation by *Hobbit* <hobbit@avian.org>.
     Rewritten with IPv6 support by Eric Jackson <ericj@monkey.org>.
(snip)
BSD                              June 25, 2001                             BSD
```

How I use nc on Docker on my macOS
```
$ docker run -it centos:7
$ yum install -y nc
$ nc -z (HAProxy) 1935
```

Here is packet captured using tcpdump:

```
# observed a problem (netmap-netcat on Linux)
12:11:50.052378 IP (An instance).52154 > (HAProxy).1935: Flags [S], seq 2390792675, win 26883, options [mss 8961,sackOK,TS val 2280100 ecr 0,nop,wscale 7], length 0
12:11:50.052808 IP (HAProxy).1935 > (An instance).52154: Flags [S.], seq 4034454280, ack 2390792676, win 26847, options [mss 8961,sackOK,TS val 110865432 ecr 2280100,nop,wscale 7], length 0
12:11:50.052825 IP (An instance).52154 > (HAProxy).1935: Flags [.], ack 1, win 211, options [nop,nop,TS val 2280100 ecr 110865432], length 0
12:11:50.052939 IP (An instance).52154 > (HAProxy).1935: Flags [F.], seq 1, ack 1, win 211, options [nop,nop,TS val 2280101 ecr 110865432], length 0
12:11:50.053489 IP (HAProxy).1935 > (An instance).52154: Flags [F.], seq 1, ack 2, win 210, options [nop,nop,TS val 110865433 ecr 2280101], length 0
12:11:50.053498 IP (An instance).52154 > (HAProxy).1935: Flags [.], ack 2, win 211, options [nop,nop,TS val 2280101 ecr 110865433], length 0


# cannot observed any problems (BSD nc -z on macOS)
$ sudo tcpdump -nn "port 1935"
21:11:58.711866 IP (macOS).51727 > (HAProxy).1935: Flags [S], seq 2549296565, win 65535, options [mss 1460,nop,wscale 6,nop,nop,TS val 493683457 ecr 0,sackOK,eol], length 0
21:11:58.730890 IP (HAProxy).1935 > (macOS).51727: Flags [S.], seq 155454489, ack 2549296566, win 26847, options [mss 1250,sackOK,TS val 110874092 ecr 493683457,nop,wscale 7], length 0
21:11:58.730933 IP (macOS).51727 > (HAProxy).1935: Flags [.], ack 1, win 2050, options [nop,nop,TS val 493683475 ecr 110874092], length 0
21:11:58.735293 IP (macOS).51727 > (HAProxy).1935: Flags [F.], seq 1, ack 1, win 2050, options [nop,nop,TS val 493683479 ecr 110874092], length 0
21:11:58.749282 IP (HAProxy).1935 > (macOS).51727: Flags [F.], seq 1, ack 2, win 210, options [nop,nop,TS val 110874110 ecr 493683479], length 0
21:11:58.749315 IP (macOS).51727 > (HAProxy).1935: Flags [.], ack 2, win 2050, options [nop,nop,TS val 493683492 ecr 110874110], length 0

# sometimes observed a problem (nmap-netcat on Docker on macOS)
$ sudo tcpdump -nn "port 1935"
21:34:15.188196 IP (macOS).52501 > (HAProxy).1935: Flags [S], seq 2433871201, win 65535, options [mss 1460,nop,wscale 6,nop,nop,TS val 494999749 ecr 0,sackOK,eol], length 0
21:34:15.237759 IP (HAProxy).1935 > (macOS).52501: Flags [S.], seq 3115702925, ack 2433871202, win 26847, options [mss 1250,sackOK,TS val 112210574 ecr 494999749,nop,wscale 7], length 0
21:34:15.237794 IP (macOS).52501 > (HAProxy).1935: Flags [.], ack 1, win 2050, options [nop,nop,TS val 494999798 ecr 112210574], length 0
21:34:15.241498 IP (macOS).52501 > (HAProxy).1935: Flags [F.], seq 1, ack 1, win 2050, options [nop,nop,TS val 494999801 ecr 112210574], length 0
21:34:15.294944 IP (HAProxy).1935 > (macOS).52501: Flags [F.], seq 1, ack 2, win 210, options [nop,nop,TS val 112210624 ecr 494999801], length 0
21:34:15.294987 IP (macOS).52501 > (HAProxy).1935: Flags [.], ack 2, win 2050, options [nop,nop,TS val 494999854 ecr 112210624], length 0
```

## Do you have an idea how to solve the issue?

Sadly no.

Thanks in advance,

**Comment by wtarreau** [comment]
Hi, sorry for the delay.

Does it fail if you allow more than one entry ? I'm asking because I suspect that we might be facing a blurry limit on the number of entries, like allocating the new entry before checking for limits to ensure that at least one succeeds in case of multi-threading. And my suspicion is then that depending how far the outgoing connection step goes during the close, it will or will not reach the table update. You mentioned "nopurge" not being set, but that's precisely the situation which allows recycling when the table is full. If it works for you with a few more entries, what I would recommend would be to set the number of entries to 1 + number of threads to have enough margin for all situations.

**Comment by kyontan** [comment]
Hi, sorry for a long delay!

> Does it fail if you allow more than one entry

Nope, it does not fail when with more than one entry.
As you said, I thought its situation relates to some blurry effects.
In this case, I have workaround that setting table size more than one, but there might be scenario that must set table size to one.

**Comment by capflam** [comment]
@EmericBr, what is the status of this issue ?

**Comment by wtarreau** [comment]
Finally found the cause! It had nothing to do with stick-tables but with the fact that the connection is aborted too early, doesn't enable the response analysers which are responsible for committing the entry! It's just been merged into 2.6-dev and will progressively be backported if no issue is discovered with it.

**Comment by wtarreau** [comment]
Let's be reasonable and avoid to backport this further than 2.4, who knows what it could break there...
